### ISSUE TITLE
Use a Credo plugin defined in the analyzed project

### ISSUE THREAD

**Issue by hochata** [rrrene/credo#990]
Hi! I'm not sure if this is expected behavior. If it is then I'll move it to the proposals repository.

### Environment

* Credo version (`mix credo -v`): 1.6.4
* Erlang/Elixir version (`elixir -v`): Elixir 1.10.0 (compiled with Erlang/OTP 21)
* Operating system: Linux

### What were you trying to do?
Use a Credo Plugin on itself. That is, add `plugins: [{CredoPlugin, []}]` on the `.credo.exs` of the project where `CredoPlugin` is defined.

### Expected outcome
Have all the functionalities defined in the Plugin.

### Actual outcome
```shell
** (credo) Plugin module `CredoPlugin` is not available and could not be initialized
```

Which is understandable, because the modules from project under analysis are not loaded by default.

I thought this could be solved by adding the files where the module is defined to the `requires: ` field, but it doesn't change anything. After a quick look, the plugins are loaded before the files from the `requires` field are loaded.

Is there a way to use a Credo Plugin defined in the project under analysis?

**Comment by rrrene** [comment]
Shouldn't this be possible by loading the current application first?

Something along the lines of  `mix do app.config, credo`?

**Comment by hochata** [comment]
Thanks for the answer! Something like this in the project's `mix.exs` works.

```elixir
  def project do
    [
      ...
      aliases: aliases()
    ]
  end

  defp credo_plugin_in_itself(_) do
    Mix.Task.run("compile")
    Mix.Task.run("credo", ["extra args for your plugin"])
  end

  defp aliases do
    [
      credo_plug_in_itself: &credo_plugin_in_itself/1
    ]
  end
```

Compiling the whole project is required to generate the `*.app` files that are needed to load the application.

Should this be documented somewhere?

**Comment by hochata** [comment]
A while it does work, it feels more like a workaround. Do you think it would be too problematic to make the `requires` field work? Or some other simpler mechanism.

**Comment by rrrene** [comment]
This feels like a workaround, because it IS a workaround. 😉 

1. Credo is a static analysis tool and because of this, it is not supposed to compile the code it is analysing.
2. Credo Plugins are dependencies loaded alongside Credo.

You can put a plugin inside the project it is supposed to analyse, but that violates both 1 and 2 - and is therefore a workaround.

P.S. Yes, It would be problematic to change the point in time when certain parts of the config are loaded while others aren't (unless I am misunderstanding the question?).

P.P.S. Wouldn't it be way easier to alias your mix task this way:

```elixir
  def project do
    [
      ...
      aliases: aliases()
    ]
  end

  defp aliases do
    [
      credo_plug_in_itself: ["compile", "credo extra args for your plugin"]
    ]
  end
```

**Comment by hochata** [comment]
Understandable, thanks for helping getting this working.

And I meant the extend the `requires` configuration option to be able to load plugin, as it currently only works for checks only. But I understand the decision to make if that way. Custom checks are project specific, while Plugins are more reusable and should be separate packages. 

Thanks! Your snippet is simpler.
